### ISSUE TITLE
Introspecting local functions, lambdas, anonymous functions and local variables is not yet fully supported in Kotlin reflection

### ISSUE THREAD

**Issue by berserkore** [aPureBase/KGraphQL#114]
Hi There,

I am struggling solving this issue. I am running the following tutorial:
https://kgraphql.io/Tutorials/starwars/
I am using the version: v0.15.6

Here is the 

> stacktrace:--------- beginning of crash
    **kotlin.reflect.jvm.internal.KotlinReflectionInternalError: Introspecting local functions, lambdas, anonymous functions and local variables is not yet fully supported in Kotlin reflection**
        at kotlin.reflect.jvm.internal.EmptyContainerForLocal.fail(EmptyContainerForLocal.kt:41)
        at kotlin.reflect.jvm.internal.EmptyContainerForLocal.getFunctions(EmptyContainerForLocal.kt:37)
        at kotlin.reflect.jvm.internal.KDeclarationContainerImpl.findFunctionDescriptor(KDeclarationContainerImpl.kt:151)
        at kotlin.reflect.jvm.internal.KFunctionImpl$descriptor$2.invoke(KFunctionImpl.kt:56)
        at kotlin.reflect.jvm.internal.KFunctionImpl$descriptor$2.invoke(KFunctionImpl.kt:36)
        at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:92)
        at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:31)
        at kotlin.reflect.jvm.internal.KFunctionImpl.getDescriptor(KFunctionImpl.kt)
        at kotlin.reflect.jvm.internal.KFunctionImpl.getDescriptor(KFunctionImpl.kt:36)
        at kotlin.reflect.jvm.internal.KCallableImpl$_returnType$1.invoke(KCallableImpl.kt:74)
        at kotlin.reflect.jvm.internal.KCallableImpl$_returnType$1.invoke(KCallableImpl.kt:21)
        at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:92)
        at kotlin.reflect.jvm.internal.KCallableImpl.getReturnType(KCallableImpl.kt:80)
        at com.apurebase.kgraphql.schema.structure.SchemaCompilation.handleOperation(**SchemaCompilation.kt:150**)
        at com.apurebase.kgraphql.schema.structure.SchemaCompilation.handleQueries(SchemaCompilation.kt:126)
        at com.apurebase.kgraphql.schema.structure.SchemaCompilation.perform(SchemaCompilation.kt:63)
        at com.apurebase.kgraphql.schema.dsl.SchemaBuilder$build$1.invokeSuspend(SchemaBuilder.kt:32)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at com.apurebase.kgraphql.schema.dsl.SchemaBuilder.build(SchemaBuilder.kt:31)
        at com.apurebase.kgraphql.KGraphQL$Companion.schema(**KGraphQL.kt:12**)

Any suggestions of fixes?

Thanks

**Comment by berserkore** [comment]
More info, it seems to be happening in the handleQueries, this issue is not happening all the time, just at random times. Please see my 2 screenshots, one successful, and the other one when it fails on the same machine/same code (running the github example):

**When it works**
<img width="1054" alt="1" src="https://user-images.githubusercontent.com/11094066/92423924-2ed8cc00-f1c6-11ea-983a-527895b4d03d.png">

**When it fails at random times** 
<img width="1036" alt="2" src="https://user-images.githubusercontent.com/11094066/92424125-c76f4c00-f1c6-11ea-8b85-d93e9b0ab371.png">
<img width="1033" alt="3" src="https://user-images.githubusercontent.com/11094066/92424121-c3dbc500-f1c6-11ea-95dd-e639f6fd2e55.png">

it is linked to the lazy kFunction variable found in FunctionWrapper 

```
override val kFunction: KFunction<T> by lazy  {implementation.reflect()!!}
```

Any ideas?

**Comment by berserkore** [comment]
Here is an update:
**It is being garbage collected on the process** which cause the kfunction reference to be lost. This is very unpredictable when the GC is going to kick in. That also explain why I get crashes intermittently I instantiate the schema. Here is pictures where I spotted GC killing reflection instances:

<img width="1035" alt="Screen Shot 2020-09-08 at 16 11 29" src="https://user-images.githubusercontent.com/11094066/92439963-14ffaf00-f1ef-11ea-8e0f-ac80cd4af555.png">
<img width="1136" alt="Screen Shot 2020-09-08 at 16 11 57" src="https://user-images.githubusercontent.com/11094066/92439975-192bcc80-f1ef-11ea-90c2-dc516d94cd9b.png">

**We need to had some hard references in several places while instantiating the schema!**

**Comment by jeggy** [comment]
Hi @berserkore, thanks a bunch for your investigation here :+1: 

Could you provide some information about your environment? Kotlin version, JVM version, which OS.

_Other than this. We are in the process of upgrading this library to Kotlin 1.4 and after this we will re-do some of the internals to depend more on the new `typeOf<T>()` functionality to have the internals of this library more stable and simpler._

**Comment by berserkore** [comment]
Hi @jeggy, I am using android 1.3.71 (same as this library) and I am running on Android OS :).
Great library by the way!

**Comment by berserkore** [comment]
Hi @jeggy, I finally found a fix for this problem. As pointed above, it was because of the garbage collector which is wiping out several allocated variables (reflected) when the heap is starving memory or put at stress. 
To fix the issue, you need to prevent GC to come and wreck your entire variable schemas when it builds up. Here is a way I did in order to fix it:

_In **FunctionWrapper.kt**, change all_

`override val kFunction: KFunction<T> by lazy  {implementation.reflect()!!}`

**to**

`override val kFunction: KFunction<T>
@Synchronized
            get() {
                return implementation.reflect()!!
            }`

You should no longer have GC crashing your program and get this misleading error..

Regarding the environment, you get this issue when you run KGraphql on poor/restrain hardware spec such as  various android devices as well as IoT devices like raspberry pi.
I have checked  the fixed using Kotlin 1.3.71 and 1.4.0, it is working just fine.
 
Hope it helps,
Best regards

**Comment by jeggy** [comment]
Thanks for fixing this in #133 :+1: 

I was never able to reproduce this issue, but I trust you that we can close this now :+1: 

This is available now in version `0.17.1`
